lib/spack/docs/build_systems/makefilepackage.rst

```diff
@@ -0,0 +1,295 @@
+installed already, so you don't need to add a dependency on ``make``.
+However, if the package uses a ``GNUmakefile`` or the developers recommend
+using GNU Make, you should add a dependency on ``gmake``:
+
+.. code-block:: python
+
+   depends_on('gmake', type='build')
+
+
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+Types of Makefile packages
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Most of the work involved in packaging software that uses Makefiles
+involves overriding or replacing hard-coded variables. Many packages
+make the mistake of hard-coding compilers, usually for GCC or Intel.
+This is fine if you happen to be using that particular compiler, but
+Spack is designed to work with *any* compiler, and you need to ensure
+that this is the case.
+
+Depending on how the Makefile is designed, there are 4 common strategies
+that can be used to set or override the appropriate variables:
+
+"""""""""""""""""""""
+Environment variables
+"""""""""""""""""""""
+
+Make has multiple types of assignment operators. Some Makefiles
+use ``=`` to assign variables. The only way to override these
+variables is to edit the Makefile or override them on the command-line.
+However, Makefiles that use ``?=`` for assignment honor environment
+variables. Since Spack already sets ``CC``, ``CXX``, ``F77``, and ``FC``,
+you won't need to worry about setting these variables. If there are
+any other variables you need to set, you can do this in the ``edit``
+method:
+
+.. code-block:: python
+
+   def edit(self, spec, prefix):
+       env['PREFIX'] = prefix
+       env['BLASLIB'] = spec['blas'].libs.ld_flags
+
+
+``cbench`` is a good example of a simple package that does this, while
+``esmf`` is a good example of a more complex package.
+
+""""""""""""""""""""""
+Command-line arguments
+""""""""""""""""""""""
+
+If the Makefile ignores environment variables, the next thing to try
+is command-line arguments. You can do this by overriding the
+``build_targets`` attribute. If you don't need access to the spec,
+you can do this like so:
+
+.. code-block:: python
+
+   build_targets = ['CC=cc']
+
+
+If you do need access to the spec, you can create a property like so:
+
+.. code-block:: python
+
+   @property
+   def build_targets(self):
+       spec = self.spec
+
+       return [
+           'CC=cc',
+           'BLASLIB={0}'.format(spec['blas'].libs.ld_flags),
+       ]
+
+
+``cloverleaf`` is a good example of a package that uses this strategy.
+
+"""""""""""""
+Edit Makefile
+"""""""""""""
+
+Some Makefiles are just plain stubborn and will ignore command-line
+variables. The only way to ensure that these packages build correctly
+is to directly edit the Makefile. Spack provides a ``FileFilter`` class
+and a ``filter_file`` method to help with this. For example:
+
+.. code-block:: python
+
+   def edit(self, spec, prefix):
+       makefile = FileFilter('Makefile')
+
+       makefile.filter('CC = gcc',  'CC = cc')
+       makefile.filter('CXX = g++', 'CC = c++')
```

Comment:
It seems inefficient to process makefile once per filter.  Is there a way that multiple filters could be applied in one pass?

Also.. the variable here should probably be called `Makefile` because that's what the file is called.

Reply:
`filter` only supports a single search and replace argument at the moment.

For the variable name, see [PEP 8](https://www.python.org/dev/peps/pep-0008/#function-and-variable-names).

Reply:
The variable is technically a constant because it's set once and never changed.